File: 689-maximum-sum-of-3-non-overlapping-subarrays/maximum-sum-of-3-non-overlapping-subarrays.py
class Solution:
    def maxSumOfThreeSubarrays(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        window = [sum(nums[:k])]

        for i in range(1, n - k + 1):
            x = sum(nums[i : i + k])
            window.append(x)

        wlen = len(window)

        left = [0] * wlen
        right = [0] * wlen

        bestleft = 0
        for i in range(wlen):
            if window[i] > window[bestleft]:
                bestleft = i
            left[i] = bestleft

        bestright = wlen - 1
        for i in range(wlen - 1, -1, -1):
            if window[i] >= window[bestright]:
                bestright = i
            right[i] = bestright

        maxsum = 0
        ans = [-1, -1, -1]

        for i in range(k, wlen - k):
            l, r = left[i - k], right[i + k]

            threesum = window[l] + window[i] + window[r]
            if threesum > maxsum:
                maxsum = threesum
                ans = [l, i, r]

        return ans

File: 689-maximum-sum-of-3-non-overlapping-subarrays/maximum-sum-of-3-non-overlapping-subarrays.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSumOfThreeSubarrays = function (nums, k) {
    const n = nums.length;
    const window = [nums.slice(0, k).reduce((a, b) => a + b, 0)];

    for (let i = 1; i < n - k + 1; i++) {
        x = nums.slice(i, i + k).reduce((a, b) => a + b, 0);
        window.push(x);
    }

    const wlen = window.length;
    const left = Array(wlen).fill(0);
    const right = Array(wlen).fill(0);

    let bestleft = 0;
    for (let i = 0; i < wlen; i++) {
        if (window[i] > window[bestleft]) {
            bestleft = i;
        }
        left[i] = bestleft;
    }

    let bestright = wlen - 1;
    for (let i = wlen - 1; i >= 0; i--) {
        if (window[i] >= window[bestright]) {
            bestright = i;
        }
        right[i] = bestright;
    }

    let maxsum = 0;
    let ans = Array(3).fill(-1);

    for (let i = k; i < wlen - k; i++) {
        let l = left[i - k], r = right[i + k];
        let threesum = window[l] + window[i] + window[r];

        if (threesum > maxsum) {
            maxsum = threesum;
            ans = [l, i, r];
        }
    }

    return ans;
};